[pipelining] test pipeline_order in schedule

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127796
* __->__ #127559
* #127332
* #127084

Add a unittest to test validate the pipeline order for different `num_stages`, `num_microbatches`, `num_world_size` combinations. This doesn't actually run the schedule but just validates the ordering of microbatches processed is valid, therefore doesn't require GPUs / multiple processes.

Will add more combinations and negative tests.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

H-Huang committed:
[pipelining] test pipeline_order in schedule

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


This PR is still WIP


cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


This PR is still WIP


cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


This PR is still WIP


cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


This PR is still WIP


cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -344,7 +362,174 @@ def test_grad_with_manual_interleaved(self, ScheduleClass):
 instantiate_parametrized_tests(ScheduleTest)
 
+
+def format_pipeline_order(pipeline_order: Dict[int, List[Optional[_Action]]]):
```
Output looks like this (for num_local_stages=4 num_microbatches=8 group_size=8):

https://gist.github.com/H-Huang/2de7531dd6605b7a7eee6b95295291c9

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


Add a unittest to test validate the pipeline order for different `num_stages`, `num_microbatches`, `num_world_size` combinations. This doesn't actually run the schedule but just validates the ordering of microbatches processed is valid, therefore doesn't require GPUs / multiple processes.

Will add more combinations and negative tests.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


Add a unittest to test validate the pipeline order for different `num_stages`, `num_microbatches`, `num_world_size` combinations. This doesn't actually run the schedule but just validates the ordering of microbatches processed is valid, therefore doesn't require GPUs / multiple processes.

Will add more combinations and negative tests.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -344,7 +362,190 @@ def test_grad_with_manual_interleaved(self, ScheduleClass):
+def format_pipeline_order(pipeline_order: Dict[int, List[Optional[_Action]]]):
+    import itertools
+
+    column_labels = ["Step " + str(i) for i in range(len(pipeline_order[0]))]
```
nit: 2 changes that would make this a lot more readable
- transpose, so you have up to RANKS cols and STEPS rows, since RANKS<<STEPS
- implement __repr__ for Action or generally, format such that you use one letter + # per item (like F1 for forward microbatch1)

H-Huang replied:
good suggestions, will add. I was also planning on moving this out to its own utility and also printing this out if there is an error

H-Huang replied:
now looks like this:

![nun local stages=2 nun microbatches=8 group size=8](https://github.com/pytorch/pytorch/assets/14858254/27339157-697b-47a7-a4a4-9975b77c3597)

wconstab commented on test/distributed/pipelining/test_schedule.py:
```diff
@@ -344,7 +362,190 @@ def test_grad_with_manual_interleaved(self, ScheduleClass):
+        """
+        pipeline_order[rank] = [(computation_type, microbatch_index, stage_index), ...]
+
+        Validating that the pipeline order follows the rules:
```
this seems like a good start. I think it's a good idea to have some 'logic verification' tests for our schedules.

However, it would be more interesting if it could check the send/recv ops including batching.  Somehow we have to figure out if we are doing the right comms or not, without actually running on gpus.

H-Huang replied:
I think that when the schedule logic is determined at runtime and you have to figure out what when to call `get_fwd_recv_ops`, `get_fwd_send_ops`, etc. then checking batching is a lot more helpful.

For these new schedules I found that there wasn't really an issue with figuring out how operations were batched since the logic based on each step is straightforward. At every time step we were guaranteed to have each batched op across ranks match with each other and if they didn't it was pretty easy to tell. Usually the issue was because a stage was too early or too late in the plan, which this test did help a lot to verify as I was developing.

But yeah it is fair that the schedules are not "actually being run" with this test so we can't say for absolute certainty that it works. If we were able to check send/recv ops that might help increase the confidence, but I am pretty confident that the current `step_microbatches` implementation is good and our current tests can catch issues with it, which in turns means the plans should "just work"

H-Huang replied:
mock the forward / backward.

Check the results of get_recv ops and get_send_ops. Validate these

H-Huang replied:
https://github.com/pytorch/pytorch/issues/127965

wconstab commented on torch/distributed/pipelining/__init__.py:
```diff
@@ -14,7 +14,7 @@
     ScheduleLoopedBFS,
 )
-from .PipelineStage import ManualPipelineStage, PipelineStage
+from .PipelineStage import _PipelineStageBase, ManualPipelineStage, PipelineStage
```
accidental merge?

wconstab replied:
oh i see it was intentional.  But is it allowed?  (RE rules about _ modules and packages)  @kwen2501 

maybe you should just import this from PipelineStage directly in the test file.

H-Huang replied:
removed it and importing it from PipelineStage directly

H-Huang committed:
Update on "[pipelining] test pipeline_order in schedule"


Add a unittest to test validate the pipeline order for different `num_stages`, `num_microbatches`, `num_world_size` combinations. This doesn't actually run the schedule but just validates the ordering of microbatches processed is valid, therefore doesn't require GPUs / multiple processes.

Will add more combinations and negative tests.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -i

H-Huang commented:
@pytorchbot merge -i
